File: packages/py/src/airlock/infrastructure/events/redis_event_bus.py
"""Redis Pub/Sub event bus (ADR-0002). A dedicated subscriber connection feeds
a background listener (`run`) that dispatches to per-topic handlers."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from airlock.application.ports.event_bus import EventHandler
from airlock.domain.events import EventTopic

if TYPE_CHECKING:
    from collections.abc import Mapping

    from pydantic import BaseModel
    from redis.asyncio import Redis
    from redis.asyncio.client import PubSub


class RedisEventBus:
    def __init__(self, publisher: Redis, subscriber: Redis) -> None:
        self._publisher = publisher
        self._subscriber = subscriber
        self._handlers: dict[str, list[EventHandler]] = {}
        self._pubsub: PubSub | None = None

    async def publish(self, topic: EventTopic, event: BaseModel) -> None:
        await self._publisher.publish(topic.value, event.model_dump_json())

    async def subscribe(self, topic: EventTopic, handler: EventHandler) -> None:
        self._handlers.setdefault(topic.value, []).append(handler)
        await self._ensure_pubsub().subscribe(topic.value)

    async def run(self) -> None:
        """Listen for messages and dispatch them until the task is cancelled."""
        async for message in self._ensure_pubsub().listen():
            await self._dispatch(message)

    def _ensure_pubsub(self) -> PubSub:
        if self._pubsub is None:
            self._pubsub = self._subscriber.pubsub()
        return self._pubsub

    async def _dispatch(self, message: Mapping[str, object]) -> None:
        if message.get("type") != "message":
            return
        channel = _as_text(message["channel"])
        payload: object = json.loads(_as_text(message["data"]))
        for handler in self._handlers.get(channel, []):
            result = handler(payload)
            if result is not None:
                await result


def _as_text(value: object) -> str:
    if isinstance(value, bytes):
        return value.decode("utf-8")
    return str(value)

File: packages/ts/src/infrastructure/events/redis-event-bus.ts
import type {
  EventHandler,
  EventPublisher,
  EventSubscriber,
} from "../../application/ports/event-bus";
import type { EventTopic } from "../../domain/events";

/** The subset of a Redis client the bus uses to publish. */
export interface RedisPublishCommand {
  publish(channel: string, message: string): Promise<number>;
}

/** The subset of a Redis client the bus uses to subscribe. */
export interface RedisSubscribeCommand {
  subscribe(channel: string): Promise<unknown>;
  on(event: "message", listener: (channel: string, message: string) => void): unknown;
}

/**
 * Event bus over Redis Pub/Sub (see ADR-0002). Needs two connections: a Redis
 * connection in subscribe mode cannot issue other commands, so publishing uses a
 * separate connection. Delivery is at-most-once; the run store stays the source
 * of truth, and resume is idempotent.
 */
export class RedisEventBus implements EventPublisher, EventSubscriber {
  private readonly handlers = new Map<string, EventHandler<unknown>[]>();
  private listening = false;

  constructor(
    private readonly publisher: RedisPublishCommand,
    private readonly subscriber: RedisSubscribeCommand,
  ) {}

  async publish<TEvent>(topic: EventTopic, event: TEvent): Promise<void> {
    await this.publisher.publish(topic, JSON.stringify(event));
  }

  async subscribe<TEvent>(topic: EventTopic, handler: EventHandler<TEvent>): Promise<void> {
    this.startListening();
    const existing = this.handlers.get(topic) ?? [];
    existing.push(handler as EventHandler<unknown>);
    this.handlers.set(topic, existing);
    await this.subscriber.subscribe(topic);
  }

  private startListening(): void {
    if (this.listening) {
      return;
    }
    this.subscriber.on("message", (channel, message) => {
      void this.dispatch(channel, message);
    });
    this.listening = true;
  }

  private async dispatch(channel: string, message: string): Promise<void> {
    const handlers = this.handlers.get(channel) ?? [];
    const event: unknown = JSON.parse(message);
    for (const handler of handlers) {
      await handler(event);
    }
  }
}
